gadgets/trace_mount/program.bpf.c

```diff
@@ -0,0 +1,194 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/* Copyright (c) 2021 Hengqi Chen */
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <gadget/macros.h>
+#include <gadget/mntns_filter.h>
+#include <gadget/types.h>
+
+#define MAX_ENTRIES 10240
+#define TASK_COMM_LEN 16
+#define FS_NAME_LEN 8
+#define DATA_LEN 512
+#define PATH_MAX 4096
+
+enum op {
+	MOUNT,
+	UMOUNT,
+};
+
+struct arg {
+	__u64 ts;
+	__u64 flags;
+	const char *src;
+	const char *dest;
+	const char *fs;
+	const char *data;
+	enum op op;
+};
+
+struct event {
+	__u64 delta;
+	__u64 flags;
+	__u32 pid;
+	__u32 tid;
+	__u64 mount_ns_id;
+	__u64 timestamp;
+	int ret;
+	__u8 comm[TASK_COMM_LEN];
+	__u8 fs[FS_NAME_LEN];
+	__u8 src[PATH_MAX];
+	__u8 dest[PATH_MAX];
+	__u8 data[DATA_LEN];
+	enum op op;
+};
+
+const volatile pid_t target_pid = 0;
+
+GADGET_PARAM(target_pid);
+
+// we need this to make sure the compiler doesn't remove our struct
+const struct event *unusedevent __attribute__((unused));
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, __u32);
+	__type(value, struct arg);
+} args SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(max_entries, 1);
+	__type(key, int);
+	__type(value, struct event);
+} heap SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(value_size, sizeof(__u32));
+} events SEC(".maps");
+
+// TODO: have to use "inline" to avoid this error:
+// bpf/mountsnoop.bpf.c:41:12: error: defined with too many args
+// static int probe_entry(const char *src, const char *dest, const char *fs,
+static __always_inline int probe_entry(const char *src, const char *dest,
+				       const char *fs, __u64 flags,
+				       const char *data, enum op op)
+{
+	__u64 pid_tgid = bpf_get_current_pid_tgid();
+	__u32 pid = pid_tgid >> 32;
+	__u32 tid = (__u32)pid_tgid;
+	struct arg arg = {};
+	u64 mntns_id;
+
+	mntns_id = gadget_get_mntns_id();
+
+	if (gadget_should_discard_mntns_id(mntns_id))
+		return 0;
+
+	if (target_pid && target_pid != pid)
+		return 0;
+
+	arg.ts = bpf_ktime_get_ns();
+	arg.flags = flags;
+	arg.src = src;
+	arg.dest = dest;
+	arg.fs = fs;
+	arg.data = data;
+	arg.op = op;
+	bpf_map_update_elem(&args, &tid, &arg, BPF_ANY);
+
+	return 0;
+};
+
+static int probe_exit(void *ctx, int ret)
+{
+	__u64 pid_tgid = bpf_get_current_pid_tgid();
+	__u32 pid = pid_tgid >> 32;
+	__u32 tid = (__u32)pid_tgid;
+	struct arg *argp;
+	struct event *eventp;
+	int zero = 0;
+
+	argp = bpf_map_lookup_elem(&args, &tid);
+	if (!argp)
+		return 0;
+
+	eventp = bpf_map_lookup_elem(&heap, &zero);
+	if (!eventp)
+		return 0;
```

Comment:
As commented on the other PR. this should cleanup the entry on the `args` map.

Another solution is to swap the lookups, first lookup on `heap` (it'll never fail and is very fast), then lookup on args.